lib/eco/autolboxdetector.py

```diff
@@ -0,0 +1,435 @@
+    wrap the error into that box."""
+    def __init__(self, origparser):
+        self.op = origparser
+        self.langs = {}
+
+    def preload(self, langname):
+        if langname in self.langs:
+            return
+
+        main = lang_dict[langname]
+
+        # preload nested languages
+        for sub in main.included_langs:
+            self.langs[sub] = get_recognizer(sub)
+
+    def find_terminal(self, node):
+        while node.children:
+            node = node.children[-1]
+        if type(node.symbol) is Terminal:
+            return node.next_term
+        return None
+
+    def detect_lbox(self, errornode):
+        # Find position on stack where lbox would be valid
+        valid = []
+        for sub in self.langs:
+            lbox = MagicTerminal("<{}>".format(sub))
+            cut = len(self.op.stack) - 1
+            while cut >= 0:
+                top = self.op.stack[cut]
+                if isinstance(top, EOS):
+                    top = top.parent.children[0] # bos
+                    state = 0
+                else:
+                    state = self.op.stack[cut].state
+                # get all possible sublangs
+                element = self.op.syntaxtable.lookup(state, lbox)
+                if type(element) in [Reduce, Shift]:
+                    term = self.find_terminal(top)
+                    if type(term) is EOS:
+                        cut = cut - 1
+                        continue
+                    if term:
+                        n = term
+                        # See if we can get a valid language box using the Recogniser
+                        r = self.langs[sub]
+                        result = r.parse(n)
+                        if r.possible_ends:
+                            # Filter results and test if remaining file can be
+                            # parsed after shifting the language box
+                            for e in r.possible_ends:
+                                if e.lookup == "<ws>" or e.lookup == "<return>":
+                                    continue
+                                if (self.contains_errornode(n, e, errornode) \
+                                    and self.parse_after_lbox(lbox, e, cut)) \
+                                    or self.parse_after_lbox(lbox, e, cut, errornode):
+                                        # Either the error was solved by
+                                        # moving it into the box or a box
+                                        # was created before it, allowing
+                                        # the error to be shifted
+                                        valid.append((n, e, sub))
+                cut = cut - 1
+        if errornode.autobox is False:
+            # XXX might have to only limit certain (node, autobox) combinations to
+            # allow boxes with different content later on
+            return # don't use this node for autoboxes anymore
+        if valid:
```

Comment:
If valid is not True, can it only be `None`? If so, can this just be `errornode.autobox = None`?

Reply:
OK, probably best left as an `if`.